doc: documentation (#4)
terraform-mongodbatlas-modules/terraform-mongodbatlas-private-endpoint: modules/aws-private-link-cluster/README.md

# AWS Private Link Cluster Terraform Submodule

> This Terraform submodule set ups a [private connection](https://www.mongodb.com/docs/atlas/security-private-endpoint/#-optional--regionalized-private-endpoints-for-multi-region-sharded-clusters) to a [MongoDB Atlas Cluster](https://www.mongodb.com/resources/products/fundamentals/mongodb-cluster-setup) utilizing [Amazon Virtual Private Cloud (aws vpc)](https://docs.aws.amazon.com/vpc/latest/userguide/what-is-amazon-vpc.html). The cluster can be either single region or geosharded (multiple regions).

Reviewer: ```suggestion
This Terraform submodule sets up a [private connection](https://www.mongodb.com/docs/atlas/security-private-endpoint/#-optional--regionalized-private-endpoints-for-multi-region-sharded-clusters) to a [MongoDB Atlas Cluster](https://www.mongodb.com/resources/products/fundamentals/mongodb-cluster-setup) utilizing [Amazon Virtual Private Cloud (aws vpc)](https://docs.aws.amazon.com/vpc/latest/userguide/what-is-amazon-vpc.html). The cluster can be either single region or geosharded (multiple regions).
```

It creates the following resources:

- A MongoDB Atlas Private Link Endpoint.
- An AWS VPC Endpoint.
- A MongoDB Atlas Private Link Endpoint Service.

> You can find detailed information of the submodule's input and output variables in the [Terraform Public Registry]()

Reviewer: Question: Should there be a link in `Terraform Public Registry` here?

Author: Yes, but as the module has not been published yet I don't have the link (we are doing the same in each module, once we publish we update the readmes to use the correct sources / links and those changes will be reflected in the following version release)

- [aws-private-link-cluster-geosharded](https://github.com/terraform-mongodbatlas-modules/terraform-mongodbatlas-private-endpoint/tree/main/examples/aws-private-link-cluster-geosharded) shows how you can use the aws private link submodule to setup regionalized private endpoints for a geosharded MongoDB Atlas Cluster utilizing Amazon Virtual Private Cloud (aws vpc).

## Resources

The module creates the following resources:

| Name | Type |
|------|------|
| [mongodbatlas_privatelink_endpoint](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/privatelink_endpoint) | resource |
| [aws_vpc_endpoint](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_endpoint) | resource |
| [mongodbatlas_privatelink_endpoint_service](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/privatelink_endpoint_service) | resource |

> Refer to the [MongoDB Atlas](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs) and [AWS](https://registry.terraform.io/providers/hashicorp/aws/latest/docs) Terraform providers documentations if you want more information.

Reviewer: ```suggestion
For more information, see the [MongoDB Atlas](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs) and [AWS](https://registry.terraform.io/providers/hashicorp/aws/latest/docs) Terraform providers documentation.
```

> In the case of the geosharded cluster, you have to enable the regionalized mode for Private Link and in each module defined provide a different AWS provider with the corresponding region. This [example](https://github.com/terraform-mongodbatlas-modules/terraform-mongodbatlas-private-endpoint/tree/main/examples/aws-private-link-cluster-geosharded) shows how you can do it.

Reviewer: ```suggestion
In the case of the geosharded cluster, you have to enable the regionalized mode for Private Link and in each module defined provide a different AWS provider with the corresponding region. For an example, see [the AWS example folder](https://github.com/terraform-mongodbatlas-modules/terraform-mongodbatlas-private-endpoint/tree/main/examples/aws-private-link-cluster-geosharded).
```

- [aws-private-link-cluster-geosharded](https://github.com/terraform-mongodbatlas-modules/terraform-mongodbatlas-private-endpoint/tree/main/examples/aws-private-link-cluster-geosharded) shows how you can use the aws private link submodule to setup regionalized private endpoints for a geosharded MongoDB Atlas Cluster utilizing Amazon Virtual Private Cloud (aws vpc).

## Resources

The module creates the following resources:

| Name | Type |
|------|------|
| [mongodbatlas_privatelink_endpoint](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/privatelink_endpoint) | resource |
| [aws_vpc_endpoint](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_endpoint) | resource |
| [mongodbatlas_privatelink_endpoint_service](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/privatelink_endpoint_service) | resource |

Refer to the [MongoDB Atlas](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs) and [AWS](https://registry.terraform.io/providers/hashicorp/aws/latest/docs) Terraform providers documentations if you want more information.

## Considerations

> In the case of the geosharded cluster, you have to enable the regionalized mode for Private Link and in each module call provide a different AWS provider with a different region. This [example](https://github.com/terraform-mongodbatlas-modules/terraform-mongodbatlas-private-endpoint/tree/main/examples/aws-private-link-cluster-geosharded) shows how you can do it.

Reviewer 2: ```suggestion
In the case of the geosharded cluster, you have to enable the regionalized mode for Private Link and in each module defined provide a different AWS provider with the corresponding region. This [example](https://github.com/terraform-mongodbatlas-modules/terraform-mongodbatlas-private-endpoint/tree/main/examples/aws-private-link-cluster-geosharded) shows how you can do it.
```

(Not related to the suggestion) How does a user enable the regionalized mode for private link?

Author: using the mongodbatlas_private_endpoint_regional_mode resource. The geosharded example shows this use case

Reviewer 2: Ohh, I see. Would it be worthing calling out in this consideration explicitly? Potentially a list like the following:

- Regionalized mode for Private Link must be enabled, this can be achieved by using the mongodbatlas_private_endpoint_regional_mode resource.
- For each region that will have a private endpoint an instance of the module must be defined using a different AWS provider.

This [example](https://github.com/terraform-mongodbatlas-modules/terraform-mongodbatlas-private-endpoint/tree/main/examples/aws-private-link-cluster-geosharded) shows how you can do it.
